src/test/java/com/google/research/bleth/simulator/ObserverTest.java

```diff
@@ -0,0 +1,930 @@
+
+        Location newLocation = randomObserver.moveTo();
+
+        assertThat(newLocation).isEqualTo(zeroOnZeroCoordinate);
+    }
+
+    @Test
+    public void moveStaticObserverStayOnItsLocation() {
+        Board board = new Board(2, 2);
+        Mockito.when(simulation.getBoard()).thenReturn(board);
+        Observer staticObserver = createStaticObserverOnLocation(zeroOnZeroCoordinate);
+
+        staticObserver.move();
+
+        assertThat(staticObserver.getLocation()).isEqualTo(zeroOnZeroCoordinate);
+    }
+
+    @Test
+    public void randomObserverSurroundedByFourDirectionsStayOnItsLocation() {
+        // -----
+        // | O |
+        // -----
+        Board board = new Board(1, 1);
+        Mockito.when(simulation.getBoard()).thenReturn(board);
+        Observer randomObserver = createRandomObserverOnLocation(zeroOnZeroCoordinate);
+
+        randomObserver.move();
+
+        assertThat(randomObserver.getLocation()).isEqualTo(zeroOnZeroCoordinate);
+    }
+
+    @Test
+    public void randomObserverSurroundedByThreeDirectionsMovesUp() {
+        // -----
+        // |   |
+        // -----
+        // | O |
+        // -----
+        Board board = new Board(2, 1);
+        Mockito.when(simulation.getBoard()).thenReturn(board);
+        Observer randomObserver = createRandomObserverOnLocation(oneOnZeroCoordinate);
+
+        randomObserver.move();
+
+        assertThat(randomObserver.getLocation()).isEqualTo(zeroOnZeroCoordinate);
+    }
+
+    @Test
+    public void randomObserverSurroundedByThreeDirectionsMovesDown() {
+        // -----
+        // | O |
+        // -----
+        // |   |
+        // -----
+        Board board = new Board(2, 1);
+        Mockito.when(simulation.getBoard()).thenReturn(board);
+        Observer randomObserver = createRandomObserverOnLocation(zeroOnZeroCoordinate);
+
+        randomObserver.move();
+
+        assertThat(randomObserver.getLocation()).isEqualTo(oneOnZeroCoordinate);
+    }
+
+    @Test
+    public void randomObserverSurroundedByThreeDirectionsMovesRight() {
+        // ----------
+        // | O |    |
+        // ----------
+        Board board = new Board(1, 2);
+        Mockito.when(simulation.getBoard()).thenReturn(board);
+        Observer randomObserver = createRandomObserverOnLocation(zeroOnZeroCoordinate);
+
+        randomObserver.move();
+
+        assertThat(randomObserver.getLocation()).isEqualTo(zeroOnOneCoordinate);
+    }
+
+    @Test
+    public void randomObserverSurroundedByThreeDirectionsMovesLeft() {
+        // ----------
+        // |   | O  |
+        // ----------
+        Board board = new Board(1, 2);
+        Mockito.when(simulation.getBoard()).thenReturn(board);
+        Observer randomObserver = createRandomObserverOnLocation(zeroOnOneCoordinate);
+
+        randomObserver.move();
+
+        assertThat(randomObserver.getLocation()).isEqualTo(zeroOnZeroCoordinate);
+    }
+
+    @Test
+    public void randomObserverMoveExactlyOneStep() {
```

Comment:
If it's a random observer, in theory this test may be "flaky": some moves will be OK, some not.
Consider instead testing e.g. 1000 consecutive moves, and verifying all of them are OK. (This just reduces uncertainty, but with random - that's all we can hope for.)